tensorflow_probability/python/experimental/mcmc/streaming_reductions_test.py

```diff
@@ -0,0 +1,127 @@
+    self._is_calibrated = is_calibrated
+    self._accepts_seed = accepts_seed
+
+  def one_step(self, current_state, previous_kernel_results, seed=None):
+    if seed is not None and not self._accepts_seed:
+      raise TypeError('seed arg not accepted')
+    return current_state + 1, TestTransitionKernelResults(
+        counter_1=previous_kernel_results.counter_1 + 1,
+        counter_2=previous_kernel_results.counter_2 + 2)
+
+  def bootstrap_results(self, current_state):
+    return TestTransitionKernelResults(counter_1=0, counter_2=0)
+
+  @property
+  def is_calibrated(self):
+    return self._is_calibrated
+
+
+@test_util.test_all_tf_execution_regimes
+class StepKernelTest(test_util.TestCase):
+
+  def test_simple_operation(self):
+    fake_kernel = TestTransitionKernel()
+    final_state, kernel_results = step_kernel(
+        num_steps=2,
+        current_state=0,
+        kernel=fake_kernel,
+        return_final_kernel_results=True
+    )
+    final_state = self.evaluate(final_state)
+    self.assertEqual(final_state, 2)
+    self.assertEqual(kernel_results.counter_1, 2)
+    self.assertEqual(kernel_results.counter_2, 4)
+
+  def test_defined_pkr(self):
+    fake_kernel = TestTransitionKernel()
+    final_state, kernel_results = step_kernel(
+        num_steps=2,
+        current_state=0,
+        previous_kernel_results=TestTransitionKernelResults(2, 3),
+        kernel=fake_kernel,
+        return_final_kernel_results=True
+    )
+    final_state = self.evaluate(final_state)
+    self.assertEqual(final_state, 2)
+    self.assertEqual(kernel_results.counter_1, 4)
+    self.assertEqual(kernel_results.counter_2, 7)
+
+  @parameterized.parameters(1, 2)
+  def test_initial_states(self, init_state):
+    fake_kernel = TestTransitionKernel()
+    final_state = step_kernel(
+        num_steps=2,
+        current_state=init_state,
+        kernel=fake_kernel,
+    )
+    final_state = self.evaluate(final_state)
+    self.assertEqual(final_state, init_state + 2)
+
+  def test_seed(self):
+    fake_kernel = TestTransitionKernel()
+    final_state, kernel_results = step_kernel(
+        num_steps=2,
+        current_state=0,
+        kernel=fake_kernel,
+        return_final_kernel_results=True,
+        seed=test_util.test_seed()
+    )
+    final_state = self.evaluate(final_state)
+    self.assertEqual(final_state, 2)
+    self.assertEqual(kernel_results.counter_1, 2)
+    self.assertEqual(kernel_results.counter_2, 4)
+
+  def test_calibration_warning(self):
+    with warnings.catch_warnings(record=True) as triggered:
+      kernel = TestTransitionKernel(is_calibrated=False)
+      tfp.mcmc.sample_chain(
+          num_results=2,
+          current_state=0,
+          kernel=kernel,
+          trace_fn=lambda current_state, kernel_results: kernel_results,
+          seed=test_util.test_seed())
+    self.assertTrue(
+        any('supplied `TransitionKernel` is not calibrated.' in str(
+            warning.message) for warning in triggered))
+
```

Comment:
Add a smoke test for reproducibility, to check that seeds are handled correctly.  For this test,

- Make a new fake kernel that ignores the current state and moves to some random (but seeded) state.
- Run at least two transitions of this kernel with `step_kernel`, at least twice with the same initial seed.
- Check that both result sequences are the same, but within one sequence the samples are very different.

Report back if anything goes mysteriously wrong -- there are subtleties of platform + seed handling that we may need to discuss offline.

Reply:
A couple of interesting facts I found out for myself (i.e. tfp seeding after `sampler.sanitize_seed` doesn't work with numpy as the seeds can become negative). Hence, I simply used some arbitrary `tfp.random` mechanism.

More importantly, an elegant way of ensuring that "within one sequence the samples are very different" didn't roll off of the top of my head, probably because `step_kernel` is meant to only return the last result by design (if you traced everything this would go against its spirit and it doesn't allow that implicitly by not giving any freedom in the loop variable shape), and with a truly random state it doesn't really support warm restart. Hence, I had to do a weird for loop implementation that does redo work.